config/test.js

```diff
@@ -33,6 +36,36 @@ function parseArgs() {
   return options;
 }
 
+/**
+ * Check if we're running full CI for a PR created on a fork. In this
+ * state, Circle CI omits secure environment variables (which is good
+ * and desired). Returns a boolean indicating whether we're in this
+ * state, and if so prints an error and sets the process exit code.
+ */
+function checkForkedPrFullRun(options) {
+  if (options.mode !== "FULL") {
+    return false;
+  }
+  if (!process.env["CIRCLE_PR_NUMBER"]) {
+    // This environment variable is only set on forked PRs.
+    // https://circleci.com/docs/2.0/env-vars/#built-in-environment-variables
+    return false;
+  }
+  if (process.env["SOURCECRED_GITHUB_TOKEN"]) {
+    return false;
+  }
+  console.error(
+    [
+      "fatal: cannot run full test suite: missing credentials",
+      "Tests on forked PRs run without credentials by default. A core team ",
+      "member will sanity-check your PR and push its head commit to a branch ",
+      "on the main SourceCred repository, which will re-run these tests.",
+    ].join("\n")
+  );
+  process.exitCode = 1;
```

Comment:
Why does checking if we are on a fork trigger these side effects? It seems much clearer to move this above to line 13, and let this check just be a check.

Reply:
This function is intended to conceptually mean “ensure that we’re not on
a fork, printing a message and exiting the process if we are (and
wtherwise returning nothing)”. E.g., in Python you’d `sys.exit(1)` here.
But in Node you’re not supposed to use `process.exit` because it
hard-exits Node without waiting for other tasks to finish, which can
lead to your stdout/stderr not being flushed, among other bad things. So
the recommended practice is this contorted mess:
<https://nodejs.org/api/process.html#process_process_exit_code>

It’s a pretty terrible API, to be sure.

I can inline the `console.error` and `process.exitCode` bit into `main`
if you like, but I think that it’s valuable that `main` be short and
easily skimmable. If you prefer, we could restructure it to

```javascript
function main() {
  const options = parseArgs();
  if (checkForkedPrFullRun(options)) {
    printForkedPrFullRunErrorMessage();
    process.exitCode = 1;
    return;
  }
  // ...
}
```

but that just feels like extra code with little value to me.

No strong feelings here; what do you think?

Reply:
I have a slight preference for the refactored version you put forward, but hearing that you considered alternative approaches is good enough for me. Since this is code that will mostly be touched only by maintainers (and hopefully accumulates complexity only slowly) I'm not as concerned with how it is factored, especially in more "stylistic" cases like this. Approved.

Reply:
Okay, changed to your preferred way.